compiler/src/dotty/tools/dotc/typer/Inferencing.scala

```diff
@@ -499,28 +499,30 @@ object Inferencing {
     propagate(accu(SimpleIdentityMap.empty, tp))
   }
 
+  /** Run the transformation after dealiasing but return the original type if it was a no-op. */
+  private def derivedOnDealias(tp: Type)(transform: Type => Type)(using Context) = {
+    val dealiased = tp.dealias
+    val transformed = transform(dealiased)
+    if transformed eq dealiased then tp // return the original type, not the result of dealiasing
+    else transformed
+  }
+
   /** Replace every top-level occurrence of a wildcard type argument by
     *  a fresh skolem type. The skolem types are of the form $i.CAP, where
     *  $i is a skolem of type `scala.internal.TypeBox`, and `CAP` is its
     *  type member. See the documentation of `TypeBox` for a rationale why we do this.
     */
-  def captureWildcards(tp: Type)(using Context): Type = tp match {
+  def captureWildcards(tp: Type)(using Context): Type = derivedOnDealias(tp) {
     case tp @ AppliedType(tycon, args) if tp.hasWildcardArg =>
-      tycon.typeParams match {
-        case tparams @ ((_: Symbol) :: _) =>
-          val boundss = tparams.map(_.paramInfo.substApprox(tparams.asInstanceOf[List[TypeSymbol]], args))
-          val args1 = args.zipWithConserve(boundss) { (arg, bounds) =>
-            arg match {
-              case TypeBounds(lo, hi) =>
-                val skolem = SkolemType(defn.TypeBoxClass.typeRef.appliedTo(lo | bounds.loBound, hi & bounds.hiBound))
-                TypeRef(skolem, defn.TypeBox_CAP)
-              case arg => arg
-            }
-          }
-          tp.derivedAppliedType(tycon, args1)
-        case _ =>
-          tp
+      val tparams = tycon.typeParamSymbols
```

Comment:
This line seems to change the semantics of the original program: If `tycon` is a type lambda instead of a class, its type parameters are not symbols.

Reply:
Meaning what happens now that didn't happen before?  I'll go back and check, and maybe even just leave it as it was, but it seemed to me that I was reusing a method that contained the same logic that was here, avoiding the cast.

Reply:
Yeah, `typeParamSymbols` is almost a structural subset of the previous code here:

```scala
  /** If `self` is a generic class, its type parameter symbols, otherwise Nil */
  final def typeParamSymbols(using Context): List[TypeSymbol] = typeParams match {
    case tparams @ (_: Symbol) :: _ =>
      assert(tparams.forall(_.isInstanceOf[Symbol]))
      tparams.asInstanceOf[List[TypeSymbol]]
        // Note: Two successive calls to typeParams can yield different results here because
        // of different completion status. I.e. the first call might produce some symbols,
        // whereas the second call gives some LambdaParams. This was observed
        // for ticket0137.scala
    case _ => Nil
  }
```

So either I'm very blind or this is fine.